include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h

```diff
@@ -927,47 +1141,202 @@ struct __brick_includes
+            // __vector_store would believe that we always have a full vector length of elements due to the starting
+            // index having greater than __preferred_vector_size elements until the end of the buffer.
+            for (std::uint8_t __i = 0; __i < __elements_to_process; ++__i)
+                __rng2[__output_start + __i] = __rng1_vector[__i];
+        }
+    }
+    template <typename _IsFull>
     void
-    operator()(const _Idx __idx, const _AccessorSrc& __acc1, _AccessorDst& __acc2) const
+    operator()(_IsFull __is_full, const std::size_t __idx, const _Range1 __rng1, _Range2 __rng2) const
     {
-        __acc2[__idx] = __acc1[__size - __idx - 1];
+        if constexpr (__base_t::__can_vectorize)
+            __vector_path_impl(__is_full, __idx, __rng1, __rng2);
+        else
+            __scalar_path_impl(__is_full, __idx, __rng1, __rng2);
     }
 };
 
 //------------------------------------------------------------------------
 // rotate_copy
 //------------------------------------------------------------------------
-template <typename _Size>
-struct __rotate_copy
+template <typename _Size, typename _Range1, typename _Range2>
+struct __rotate_copy : public walk_vector_or_scalar_base<_Range1, _Range2>
 {
+  private:
+    using __base_t = walk_vector_or_scalar_base<_Range1, _Range2>;
+    using _ValueType = oneapi::dpl::__internal::__value_t<_Range1>;
     _Size __size;
     _Size __shift;
-    template <typename _Idx, typename _AccessorSrc, typename _AccessorDst>
+
+  public:
+    __rotate_copy(_Size __size, _Size __shift) : __size(__size), __shift(__shift) {}
+
+    template <typename _IsFull>
+    void
+    __vector_path_impl(_IsFull __is_full, const std::size_t __idx, const _Range1 __rng1, _Range2 __rng2) const
+    {
+        std::size_t __shifted_idx = __shift + __idx;
+        std::size_t __wrapped_idx = __shifted_idx % __size;
+        std::size_t __n = __size;
+        _ValueType __rng1_vector[__base_t::__preferred_vector_size];
+        //1. Vectorize loads only if we know the wrap around point is beyond the current vector elements to process
+        if (__wrapped_idx + __base_t::__preferred_vector_size <= __size)
+        {
+            oneapi::dpl::__par_backend_hetero::__vector_load<__base_t::__preferred_vector_size>{__n}(
+                __is_full, __wrapped_idx, oneapi::dpl::__par_backend_hetero::__scalar_load_op{}, __rng1, __rng1_vector);
+        }
+        else
+        {
+            // A single point of non-contiguity within the rotation operation. Manually process the loop here as the
+            // access pattern becomes non-vectorizable.
+            std::size_t __remaining_elements = __n - __idx;
+            std::uint8_t __elements_to_process =
+                std::min(std::size_t{__base_t::__preferred_vector_size}, __remaining_elements);
+            for (std::uint8_t __i = 0; __i != __elements_to_process; ++__i)
+                __rng1_vector[__i] = __rng1[(__shifted_idx + __i) % __size];
```

Comment:
This is kind of a nit pick because this is only run in 1 work-item where there is non-contiguity.   However, modulo is a pretty slow operation to use in the inner loop here when we could break easily this into two loops and process directly with no modulo, and that one work item will slow the rest of the workgroup down.

Reply:
Good point. I have switched to 2 loops here. The first processes the elements up until the end of the buffer and the second processes from index `0` to the remaining vector elements.